[MOD COMPATIBILITY] Alternate Equipment Loading

digita-LUNA opened the pull request:
Adds a graphic enhancement that completely reworks how link's equipment is loaded, allowing for better implementation of Equipment mods.

**NOTE: This currently doesn't change vanilla behavior, and is meant for mods only**

- Loads equipment separately from hands, letting custom links to be used alongside custom equipment easily
- Reworks how sheaths are loaded, allowing for shield and sword mods to not be connected to each other, and the Biggoron Sword to have its own sheath
- This also allows for Timeless Equipment to work seamlessly, displaying the right equipment properly, scaled to better fit in the opposite Link's hand.
- Working Version of [#2917](https://github.com/HarbourMasters/Shipwright/pull/2917)

Any previous known bug has been fixed

Previous template no longer works due to changes in what items are loaded. Now uses completely custom display lists. If these DLs aren't found, then the vanilla system is used.

### Pictures
![soh 2023-06-13 20-32-56](https://github.com/HarbourMasters/Shipwright/assets/46070717/b079537c-e12d-4cef-83da-2f8aca90152e)
![soh 2023-06-13 20-33-11](https://github.com/HarbourMasters/Shipwright/assets/46070717/5a96686e-f173-4c19-ab53-d9e7ea7f87d4)![Uploadin
![soh 2023-06-13 20-35-32](https://github.com/HarbourMasters/Shipwright/assets/46070717/8ea96f82-2c2e-4e31-a3e4-f848cfe79586)
![soh 2023-06-13 20-36-53](https://github.com/HarbourMasters/Shipwright/assets/46070717/8acc484f-b941-4480-9928-b35f276bdb92)
![soh 2023-06-13 20-33-44](https://github.com/HarbourMasters/Shipwright/assets/46070717/a5c66129-f092-4a23-81eb-5c8b6754fb91)


<!--- section:artifacts:start -->
### Build Artifacts
<!--- section:artifacts:end -->

digita-LUNA committed:
Adding comments that properly explain the nomenclature confusion with the hookshot

digita-LUNA committed:
Rename All Link DLs

digita-LUNA committed:
Changed names of global DLists

digita-LUNA committed:
Merge branch 'HarbourMasters:develop' into dlist-tidy

digita-LUNA committed:
Begin Working on Alternate Assets

digita-LUNA committed:
Merge branch 'HarbourMasters:develop' into dlist-tidy

digita-LUNA committed:
Merge remote-tracking branch 'upstream/develop' into develop-newassets

digita-LUNA committed:
Fix errors

digita-LUNA committed:
Merge branch 'dlist-tidy' into develop-newassets

digita-LUNA committed:
Finished Main Implementation

digita-LUNA committed:
Ocarinas, Code Polishing, Child Hylian Shield

digita-LUNA committed:
Merge branch 'HarbourMasters:develop' into develop-newassets

digita-LUNA committed:
Fixed bug with the Boomerang

digita-LUNA committed:
Fixed Sword Lengths being altered by resizing

digita-LUNA committed:
Fix Crawling, Normal Hylian Shield

digita-LUNA committed:
somehow didn't come with the last commit

PurpleHato reviewed (changes requested):
I have a tiny nitpick from the cvar name "gNormalHylianShield" that makes it sounds like an anomaly
Maybe something like "gLinkChildHoldHylianShield" would sounds better, yeah it's longer but more comprehensive
outside that, it looks good

digita-LUNA committed:
Majora's Mask-Styled Bottles

digita-LUNA committed:
Rename Hylian Shield Code

digita-LUNA committed:
Equipment Screen

digita-LUNA committed:
Proper arrow placement for child

jbodner09 commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -853,8 +853,18 @@ void DrawEnhancementsMenu() {
+                UIWidgets::PaddedEnhancementCheckbox("Alternate Equipment Loading", "gAltLinkEquip", true, false);
+                UIWidgets::Tooltip("Use alternate loading method for loading Link's assets");
+                UIWidgets::PaddedEnhancementCheckbox("Majora's Mask Bottles", "gMMBottles", true, false);
+                UIWidgets::Tooltip("Use alternate loading method Bottles, mimicing Majora's Mask");
```
To somebody who has no idea what these options do, these tooltips don't quite say enough about what will happen when you turn them on compared to what's happening when they're off.  For the first one, you've already got a great description at the top of this PR; just condense it here to something like "Display Link's equipment separately from his body, allowing Timeless equipment to display properly on child Link and equipment to display properly on custom Link models."  

Similarly, for the bottles, you could say "Display bottle contents separately from the bottle itself, similar to Majora's Mask."

digita-LUNA replied:
Will do. I'm also gonna take a lot of the function renaming here and pull it back to zret itself due to how much i've sifted through figuring this out.

digita-LUNA committed:
Fix shields in EQ screen

digita-LUNA committed:
More function labelling, fixed sheaths in crawlspaces

digita-LUNA committed:
Updated A lot of functions to match ZRET names

digita-LUNA committed:
updated descriptions of options

digita-LUNA committed:
Fixed more bugs

digita-LUNA committed:
Broken Sticks and Swords are no longer age-dependent

digita-LUNA committed:
cleanup, optimization, and comprehensive comments

digita-LUNA committed:
optimized Player_SetModels

digita-LUNA committed:
Merge branch 'HarbourMasters:develop' into develop-newassets

digita-LUNA committed:
optimization fix

digita-LUNA committed:
Pedestal Master Sword

digita-LUNA committed:
Merge remote-tracking branch 'upstream/develop' into develop-newassets

digita-LUNA committed:
another fix from optimization

digita-LUNA committed:
Fixed another bug with the optimizations

digita-LUNA committed:
Child Hylian Shield can now reflect Nuts and Rocks

garrettjoecox commented:
I’m not sure how git savvy you are, but having this PR condensed into a few commits, with all documentation and renames being separate in its own commit would make this much easier to review

digita-LUNA commented:
> I’m not sure how git savvy you are, but having this PR condensed into a few commits, with all documentation and renames being separate in its own commit would make this much easier to review

The main reason for the large amount of commits was because it started as a draft and needed a lot of commits to optimize code and fix bugs.

garrettjoecox commented:
> The main reason for the large amount of commits was because it started as a draft and needed a lot of commits to optimize code and fix bugs.

Yeah I figured as much, the git savvy comment maybe came off wrong, I more meant if you know how, you should try to clean it up, but if not that’s understandable cause squashing/removing merge commits can be tricky

digita-LUNA committed:
Updated to not reuse existing Display Lists
- System now uses custom Display Lists defined in soh_assets.h
- Default Models in system are blank placeholders meant to be replaced
- Removed MM Bottles, as the setting is tied into the main option. If bottle contents don't need to be a seperate mesh, the contents model can be left blank

digita-LUNA committed:
Hookshot Update
- Hookshot Tips and Chains now Have seperate meshes depending on age, as they currently cannot be easily scaled. It's possible that this could be done better in the future
- Fixed Kokiri Sword Sheath

digita-LUNA committed:
Removed redundant call to check for Ocarina

NEstelami commented:
Great work on this PR!

Which display lists are intended to have the isolated weapons? I could be missing something but I'm only seeing mentions of weapons combined with hands.

digita-LUNA commented:
> Great work on this PR!
> 
> Which display lists are intended to have the isolated weapons? I could be missing something but I'm only seeing mentions of weapons combined with hands.

Somewhat easy to miss since they're loaded in a different place now

All Extra DLs are in objects/object_link_equipment, outside of two:
- objects/object_link_boy/gLinkAdultFPSHandDL: _Adult FPS Hand_
- objects/object_link_child/gLinkChildFPSHandDL: _Child FPS Hand_

The rest are as follows:
- gLinkBoomerangDL: _Boomerang_
- gLinkBowDL: _Fairy Bow_
- gLinkFairyOcarinaDL: _Fairy Ocarina_
- gLinkHammerDL: _Megaton Hammer_
- gLinkHookshotDL: _Hookshot_
- gLinkHookshotChainDL, gLinkHookshotSmallChainDL: _Adult and Child Hookshot Chain_
- gLinkHookshotTipDL, gLinkHookshotSmallTipDL: _Adult and Child Hookshot Tip_
- gLinkOcarinaOfTimeDL: _Ocarina of Time, Based on Child OoT DL_
- gLinkSlingshotDL: _Fairy Slingshot_

**Swords are Ordered with their model in hand, sheathed, and the sheath itself.**
- gLinkKokiriSwordDL, gLinkKokiriSwordInSheathDL, gLinkKokiriSwordSheathDL: _Kokiri Sword_
- gLinkMasterSwordDL, gLinkMasterSwordInSheathDL, gLinkMasterSwordSheathDL: _Master Sword_
- gLinkLongswordDL, gLinkLongswordInSheathDL, gLinkLongswordSheathDL: _Giant's Knife/Biggoron Sword_
- gLinkBrokenLongswordDL: _Broken Giant's Knife_

**Each Shield is ordered in models in hand, and model on the player's back**
- gLinkDekuShieldDL, gLinkDekuShieldOnBackDL: _Deku Shield_
- gLinkHylianShieldDL, gLinkHylianShieldOnBackDL: _Hylian Shield_
- gLinkMirrorShieldDL, gLinkMirrorShieldOnBackDL: _Mirror Shield_
- gLinkHylianShieldOnChildBackDL: _Child Hylian Shield_
- gLinkBottleDL: _Bottle. Adult and Child bottle DLs are almost identical in base game, so using a single bottle won't cause too much inaccuracy for a custom bottle._
- gLinkBottleContentsDL: _MM-Style Bottle contents. Only renders if the bottle isn't empty, leave as is (blank placeholder) if there doesn't need to be a separate model._

PurpleHato commented:
> * gLinkBottleDL: Bottle. Adult and Child bottle DLs are almost identical in base game, so using a single bottle won't cause too much inaccuracy for a custom bottle.

From what I understand, you mean that Adult and Child have a bottle model that is different and you decied to use only one of them for both ages, for accuracy reason, I would sugget that to be changed to use their original counterpart, even as tiny as it is, we still need to provide the game in it's original visual looking side

digita-LUNA commented:
> > * gLinkBottleDL: Bottle. Adult and Child bottle DLs are almost identical in base game, so using a single bottle won't cause too much inaccuracy for a custom bottle.
> 
> From what I understand, you mean that Adult and Child have a bottle model that is different and you decied to use only one of them for both ages, for accuracy reason, I would sugget that to be changed to use their original counterpart, even as tiny as it is, we still need to provide the game in it's original visual looking side

I understand your point, but this is meant for model mods only, as making it the standard system would require editing game assets, and that's not possible right now, as far as I know. The model for the adult and child bottle are **identical**, with the only difference in the Display Lists being an extremely slight difference in position, seen below
![image](https://github.com/HarbourMasters/Shipwright/assets/46070717/812b5653-95e9-4d2a-a530-dde96b519ca2)
While, yes, it's still technically off from vanilla behavior, I don't see the need to make a separate model entirely, at most, maybe slightly adjust it for adult with an in-engine translation like I did with the ocarina of time.

NEstelami commented on soh/src/code/z_player_lib.c:
```diff
@@ -1615,24 +1991,398 @@ void func_80090D20(PlayState* play, s32 limbIndex, Gfx** dList, Vec3s* rot, void
+                            break;
+                        case PLAYER_SHIELD_HYLIAN:
+                            if (!Player_IsChildWithHylianShield(this)) {
+                                OPEN_DISPS(play->state.gfxCtx);
+
+                                if (LINK_IS_CHILD) {
+                                    Matrix_Scale(.80f, .80f, .80f, MTXMODE_APPLY);
+                                }
+
+                                gSPMatrix(POLY_OPA_DISP++, MATRIX_NEWMTX(play->state.gfxCtx),
+                                          G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+                                gSPDisplayList(POLY_OPA_DISP++, (Gfx*)gLinkHylianShieldOnBackDL);
```
I'm noticing this pattern of function calls seems to be in a lot of places. In order to make things a bit cleaner, could you make a function like `Player_DrawItemChildScaled(float childScale, Gfx* dlist)` that contains this block? So for the hylian shield here this block would be condensed down to just `Player_DrawItemChildScaled(0.80f, gLinkHylianShieldOnBackDL);`.

digita-LUNA replied:
Updated that. The scale isn't an input, but the scale is now the proper scale for adult to child and back, 1.35 instead of 1.25, though it's not as easily inversible.

digita-LUNA committed:
Further Optimized Code + Bugfix
Fixed Bottles rendering default OoT bottles when empty

digita-LUNA committed:
fixed ANOTHER BUG i think

digita-LUNA committed:
Dark Link no longer erroneously loads weapons

digita-LUNA committed:
Fixed the Modded Loading disabling is alternate assets are disabled

digita-LUNA committed:
Merge branch 'HarbourMasters:develop' into develop-newassets

digita-LUNA committed:
removed (most) placeholder models
instead, when a custom DL isn't present, the vanilla system is used
reworks how sheaths load entirely, shouldn't affect the game without cheats

digita-LUNA committed:
Make it so that Link's bow SHOULDN'T draw if Link's in FPS mode and he doesn't have an arm

digita-LUNA committed:
my ass forgot to declare a global function

digita-LUNA committed:
oops, accidentally deleted something

PurpleHato commented:
I'm checking up on this pull request to identify any remaining obstacles, if there are any or If assistance is needed, Might be cool to have if we can move towards a review soon. The "blocking" tag has been there for about three weeks
Getting this reviewed sooner would really enhance the modding capabilities and address the current limitations we're facing :D

PurpleHato reviewed (changes requested):
~~After some careful consideration and discussions with @NEstelami , we've come to a mutual agreement that placing this system under a Cvar, rather than relying on model detection, would be more appropriate. This approach ensures that the creativity of modders isn't restricted to just one set of hands or weapons.~~

~~Therefore, individuals who are inclined towards an advanced mode, which involves utilizing multiple hands or specific gear types in certain cases, will continue to utilize the "previous" system. Meanwhile, those who are seeking simplicity in their approach will be able to make use of the new system.~~

EDIT: DISREGARD, I finally understood how that things work, great job!

digita-LUNA commented:
> After some careful consideration and discussions with @NEstelami , we've come to a mutual agreement that placing this system under a Cvar, rather than relying on model detection, would be more appropriate. This approach ensures that the creativity of modders isn't restricted to just one set of hands or weapons.
> 
> Therefore, individuals who are inclined towards an advanced mode, which involves utilizing multiple hands or specific gear types in certain cases, will continue to utilize the "previous" system. Meanwhile, those who are seeking simplicity in their approach will be able to make use of the new system.

Not sure what this means. When based on model detection, it still can utilize the previous system. I'm not entirely sure what you mean by that.

PurpleHato commented:
> > After some careful consideration and discussions with @NEstelami , we've come to a mutual agreement that placing this system under a Cvar, rather than relying on model detection, would be more appropriate. This approach ensures that the creativity of modders isn't restricted to just one set of hands or weapons.
> > Therefore, individuals who are inclined towards an advanced mode, which involves utilizing multiple hands or specific gear types in certain cases, will continue to utilize the "previous" system. Meanwhile, those who are seeking simplicity in their approach will be able to make use of the new system.
> 
> Not sure what this means. When based on model detection, it still can utilize the previous system. I'm not entirely sure what you mean by that.

~~Let's say, the user want the open hand on Shield 2, and a closed hand on Shield 3 and also want to make Shield 2 invisible on the back but not when holding, while Shield 3 would be a different model in both cases
How would they do that ?~~

EDIT: DISREGARD, I finally understood how that things work, great job!

briaguya0 commented on soh/src/code/z_parameter.c:
```diff
@@ -1851,9 +1851,12 @@ u8 Item_Give(PlayState* play, u8 item) {
             }
             
         } else if (item == ITEM_SWORD_MASTER) {
+            //this forces the sword to the Master Sword, despite the fact I don't think this is needed. Breaks DList Rework, but messes with rando / timeless equipment?
+            /*
             gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
             gSaveContext.equips.equipment &= 0xFFF0;
             gSaveContext.equips.equipment |= 0x0002;
+            */
```
it seems these lines have been around since the repo went public https://github.com/HarbourMasters/Shipwright/blame/develop/soh/src/code/z_parameter.c#L1854

digita-LUNA replied:
I'd assume so. My guess is that it's (redundant) vanilla behavior. Potentially a holdover from an earlier build of the game when it WAS needed.

Archez replied:
Without a proper decomp of a non-debug rom, we cannot make assumptions on what is a "holdover". Removing this is changing authentic behavior of the game.

In this case, this block is responsible for setting the master sword as your active sword when picking it up during the Ganon fight. We do not want to remove this.

digita-LUNA replied:
The problem is that this **also** forces the sword to the Master Sword as Young Link draws the master sword, forcing his sheath to the MS sheath.

digita-LUNA replied:
At the very least, I've uncommented the code to make sure the ganon fight works

digita-LUNA replied:
Should be fixed now.

digita-LUNA committed:
Fixed Gannon Fight

digita-LUNA committed:
Created a more proper fix for the visual bug related to the pedestal of time

sheepytina commented on soh/src/code/z_parameter.c:
```diff
@@ -1851,9 +1851,17 @@ u8 Item_Give(PlayState* play, u8 item) {
+            - Drawing the Master Sword from its pedestal
+            - Fighting Ganon, after the first phase
+            Breaks DList Rework for drawing the sword, needed for Gannon Fight.
+            To fix this, only run this code as when Link is an adult, meaning Link MUST be fighting Ganon*/
```
I know this has been a sore spot, but this really isn't a proper fix. 

Maybe you should actually be checking if the alternative loading behavior is active or not. 
`if` it isn't, `then` use vanilla behavior. 
`else` it is, `then` run alternative code that doesn't interfere with vanilla behavior.

digita-LUNA replied:
alternate loading detects to see if the proper model can use the new system.

Pepe20129 replied:
If it's preferred it could be changed like this:
```diff
-if (LINK_IS_ADULT)
+if (!ResourceGetIsCustomByName(gLinkMasterSwordDL) || LINK_IS_ADULT)
```
This way the code would be executed only if link is adult when there's a custom master sword dl and it would always be executed when there isn't one.

I think the current way it's fine but some may prefer this to retain vanilla behaviour when there's no custom dl.

inspectredc replied:
I'm definitely on the side that some other fix needs to be found to solve this issue. I haven't fully understood in past conversations what the exact result of the problem this section of code causes, but if it's a matter of being child link and having master sword without timeless equips on, then there are other methods for this to occur and would run into the same problems (or possibly to be in a situation where this path is called as child when fighting ganon). In a section where the code is granting an item to the player, I do not think that a change should be made in order to accommodate for custom dl loading.

sheepytina commented on soh/src/overlays/effects/ovl_Effect_Ss_Stick/z_eff_ss_stick.c:
```diff
@@ -50,20 +52,21 @@ u32 EffectSsStick_Init(PlayState* play, u32 index, EffectSs* this, void* initPar
-
-    if (!LINK_IS_ADULT) {
+    
+    if (Player_HoldsStick(player)) {
```
_(I should preface this by saying I'm not 100% clear on how the Deku Stick code works or why you've changed this.)_

I have similar concerns here too. Especially since [an older commit](https://github.com/HarbourMasters/Shipwright/pull/3002/commits/b72c1682f50d2fae10fc9f78dc9060ff4f363e88#diff-c071dd0ce6a1a1b51bde4fbcfccd91871c40a08457c4e99c535a8cccf17b4c8dR65) did have you separating its function into whether or not Alternate Equipment Loading is activated, before you changed how checking for this worked.

Should this maybe be using the Timeless Equipment CVar in the check?
```suggestion
    if (!LINK_IS_ADULT || (CVarGetInteger("gTimelessEquipment", 0) && Player_HoldsStick(player))) {
```
If I'm not mistaken, this way it will use the vanilla condition if the CVar is off, and allow the alternative behavior as Adult Link only if the CVar is on and the alternative condition is met.

sheepytina replied:
(Oh, and, similarly in the `EffectSsStick_Init` function.)
```c
    Player* player = GET_PLAYER(play);
    if (CVarGetInteger("gTimelessEquipment", 0))
        ageInfoEntry = Player_HoldsStick(player) + drawInfo;
```

sheepytina committed:
Retain vanilla behavior in z_eff_ss_stick.c

by checking the alternative behavior against console variable "gTimelessEquipment".

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -332,12 +395,248 @@ void Player_SetBootData(PlayState* play, Player* this) {
+
+uint8_t Player_CanUseNewLoadingMethodLeftHand(Player * this) {
+    switch (this->leftHandType) {
+        case 2: // Unused, but a safe measure
```
If this part is unused, I feel as though it should either break or not be included?

Would also be preferable for this section to have enums for describing the leftHandType cases, as the magic numbers don't make it clear what exactly is going on

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -369,18 +664,20 @@ s32 Player_ActionToModelGroup(Player* this, s32 actionParam) {
                 this->sheathType = 17;
             }
-            this->sheathDLists = &sPlayerDListGroups[this->sheathType][gSaveContext.linkAge];
+            this->sheathDLists = &sPlayerDListGroupsAlt[this->sheathType][gSaveContext.linkAge];
```
More of a question here to help my understanding; does this mean that the alt DList always going to be used? if so, is that the intention and why is it required instead of using a Player_CanUseNewLoadingMethod?

inspectredc reviewed (changes requested):
Leaving a few comments here, overall this looks very cool and I'm loving a lot of the added documentation that comes with this.

I think I will need to have a deeper look into the z_player_lib at a later time as I am not fully convinced right now from looking at the code that certain cases are being handled appropriately, e.g. ocarinas, but the general checking against handtype sections look good.

My only other nitpick would be to always use braces for if statements as I think it just generally helps a lot with readability

sheepytina committed:
Merge branch 'MoriyaFaith-newassets' into MoriyaFaith-newassets-branch3

sheepytina committed:
Completely revert all changes made to z_parameter.c

sheepytina committed:
(z_player_lib.c) Fix merge errors and tidy up comments, and add some documentation for the new functions.

I need all the help I can get to understand this code. Please forgive me for the verbose comments.

sheepytina committed:
(overlays) Formatting.

+ Replace one of Luna's comments in z_player.c with a TODO reminder to check out if what she was flagging was still, or ever, an issue.

sheepytina committed:
(z_parameter.c) Add a TODO comment, about *that* issue.

I have no idea what to do about this, so please help me out.

sheepytina committed:
(z_player_lib.c) Add previous code review comments as TODOs.

sheepytina committed:
More formatting.

sheepytina committed:
Reinstate "gAltLinkEquip" cvar, now as part of the CanUse functions.

Adds cvar checks for "gAltLinkEquip" to the `Player_CanUseNewLoadingMethod` family of functions that were once intended to replace it.

+ (z_player_lib.c) Partial formatting pass.
+ Repurpose the ghost instances of checking against gAltLinkEquip in z_arms_hook.c that Luna didn't get around to removing before. (This may need further attention.)

Explanation:

The "gAltLinkEquip" cvar was once in this repo but replaced with the `Player_CanUseNewLoadingMethod` family of functions. (see commit `51f4fa24`)
The original author's intent was to make this seamless to the player without altering vanilla behavior. For a variety of reasons this doesn't really bear out, and as I understand it, a cvar thusly still needs to be present to be able to override it.

Unlike before, this cvar now defaults to true and is otherwise an invisible option to the general player.
If turned off in the developer console, the game should fall back on vanilla behavior. At least for these aspects.

This hopefully goes some way to addressing the issue of Alternate Equipment Loading not being clearly separate from vanilla behavior, while also keeping the intent to make using mods that require this functionality seamless for players.

sheepytina committed:
Block out more modified behavior behind cvar checks.

sheepytina committed:
I found this one confusing. Please review.

sheepytina committed:
Block out Alt Equipment behavior in the following functions.

Player_SetModelsForHoldingShield
Player_SetModels

sheepytina committed:
Revert "I found this one confusing. Please review."

This reverts commit 10ffec28a257017e27aab94ee2c90d5da5edba1f.

sheepytina committed:
Small tweak to one of the cvar checks I added earlier.

sheepytina committed:
Change "alternative" comments to "alternate" for consistency

sheepytina committed:
Tweak `Player_IsChildWithHylianShield` function.

I actually didn't realise this function wasn't a new one.

sheepytina committed:
(z_player_lib.c) Tidying up some cvar checks + comments.

+ Undo some irrelevant formatting changes to make the pull request more readable.

sheepytina committed:
Remove placeholder function, and revert to cvar check.

Removes `Player_CanUseNewLoadingMethod` function.

This aspect of the code probably wasn't finished, as this function didn't actually do anything and was only used in one place.

sheepytina committed:
Comments

for clarity and consistency.

sheepytina committed:
Comments about hands

(We really do need enums for leftHandType and rightHandType.)

sheepytina committed:
Fix for crash related to alternate Hookshot tip and chain.

Fix a crash where toggling Alternate Assets while holding the Hookshot/Longshot will cause the game to try and render a DList that doesn't exist.

In future I want to rework the Hookshot and come up with an actual solution for why these models couldn't be scaled.

sheepytina committed:
Slightly better fix for Hookshot tip/chain crash.

+ Remove original comments relating to the crash.

sheepytina committed:
Tweak comments and tidy things up.

sheepytina committed:
Fix (harmless) oversight with Hookshot + some comments

sheepytina committed:
Use player enums for left and right hands.

For code specific to this branch.

I've avoided touching anything that'll be merged by https://github.com/HarbourMasters/Shipwright/pull/3281

sheepytina committed:
Simplify Hookshot effects model check function.

sheepytina committed:
Whitespace, comments, and a tweak to `Player_SetModels`

inspectredc committed:
remove asset changes

inspectredc committed:
Merge branch 'develop' into custom-equipment-develop

inspectredc committed:
remove gNormalChildHylianShield

inspectredc committed:
remove timeless equipment stick stuff

inspectredc committed:
remove left hand type 20 stuff for master sword pedestal

inspectredc committed:
Separate new functions to AlternateEquipment

inspectredc committed:
remove commented out funcs

inspectredc committed:
remove scattered dlist stuff

inspectredc committed:
implement dlist swap

inspectredc committed:
test windows fix

inspectredc committed:
test 2

inspectredc committed:
build on windows

digita-LUNA committed:
Merge remote-tracking branch 'upstream/develop' into develop-newassets

digita-LUNA committed:
Update to develop

digita-LUNA committed:
i'm trying to fix things okay

digita-LUNA committed:
let it work this time

digita-LUNA committed:
Merge remote-tracking branch 'inspect/custom-equipment-develop' into develop-newassets

digita-LUNA commented:
Closed because not only is this branch completely broken, but this PR will be split into multiple smaller PRs to make pulling easier.
